### PR TITLE
feature: make runtime choosing supported in CRI managers for Kubernetes

### LINKED ISSUE #1580: [feature request] make runtime advantages support in CRI managers for Kubernetes
### Ⅰ. Issue Description

Currently, PouchContainer has lots of advantages over other container runtimes, such as:

* resource review isolation via lxcfs, 
* diskquota for multi-directory in container, 
* runtime choosing for runc-based container or runv-based containers,
* rich container mode 
* and so on

While actually in kubernetes, there is no related API to support these feature. 

While for these extended features in Kubernetes, Kubernetes has a hiden way to support this: **make user-defined parameters in annotations field in pod's definition**.

However, in order to implement this, I think we have to support these parameters in annotations in CRI manager(CRI implementation of PouchContainer).

When CRI manager deals the annotation details, it could pass these parameters to container manager, and container manager definitely implement these features very well.

So, we need to accomplish the following things:

* define the specific naming in annotations for each feature;
* implement the transformation in CRI manager and pass them to container manager.

### PR DESCRIPTION
**Pull request by Starnop**
Signed-off-by: Starnop <starnop@163.com>

### Ⅱ. Does this pull request fix one issue?

fixes #1580

### Ⅳ. Describe how to verify it
1. Prerequisites Installation, the runtime binaries you will use is necessary
2. You should start pouchd with the configuration like 
```
pouchd --enable-cri --cri-version v1alpha2 --add-runtime runv=runv >pouchd.log 2>&1  &
```
3. After setting up your kubernetes cluster, you can create a deployment like this :
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: pouch
  labels:
    pouch: pouch
spec:
  selector:
    matchLabels:
      pouch: pouch
  template:
    metadata:
      labels:
        pouch: pouch
      annotations:
        io.kubernetes.runtime: runv
    spec:
      containers:
      - name: pouch
        image: docker.io/library/nginx:latest
        ports:
        - containerPort: 80
```
3. Use command  `pouch ps` to observe the runtime of the container.

### CODE REVIEW

File: daemon/mgr/container.go [hunk 1]
```diff
@@ -248,9 +252,12 @@ func (mgr *ContainerManager) Create(ctx context.Context, name string, config *ty
 		return nil, fmt.Errorf("NetworkingConfig cannot be nil")
 	}
 
-	id, err := mgr.generateID()
-	if err != nil {
-		return nil, err
+	id := ""
+	if id = config.ContainerID; id == "" {
```
Comment (YaoZengzeng) [inline daemon/mgr/container.go]: Why not code like this:
```go
id := config.ContainerID
if id == "" {
...
}
```
It seems more concise. 😄

Comment (Starnop) [reply daemon/mgr/container.go]: Done.

File: cri/v1alpha2/cri_utils.go [hunk 2]
```diff
@@ -235,18 +236,29 @@ func applySandboxLinuxOptions(hc *apitypes.HostConfig, lc *runtime.LinuxPodSandb
 }
 
 // makeSandboxPouchConfig returns apitypes.ContainerCreateConfig based on runtime.PodSandboxConfig.
-func makeSandboxPouchConfig(config *runtime.PodSandboxConfig, image string) (*apitypes.ContainerCreateConfig, error) {
+func makeSandboxPouchConfig(config *runtime.PodSandboxConfig, sandboxID string, image string) (*apitypes.ContainerCreateConfig, error) {
```
Comment (shaloulcy) [inline cri/v1alpha2/cri_utils.go]: sandboxID, image string

File: daemon/annotations/annotations.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package annotations
```
Comment (shaloulcy) [inline daemon/annotations/annotations.go]: maybe we should put the annotations into cri package

File: cri/v1alpha2/cri_utils.go (hunk 2 above)
Comment (Starnop) [reply cri/v1alpha2/cri_utils.go]: Done.

File: cri/v1alpha1/cri.go [hunk 4]
```diff
@@ -161,19 +162,23 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		return nil, err
 	}
 
+	id, err := c.ContainerMgr.GenerateID()
```
[the remark this answers is not in the document]
Comment (Starnop) [reply cri/v1alpha1/cri.go]: It will be used before sandbox has been created.

File: apis/types/container_config.go [hunk 5]
```diff
@@ -35,6 +35,9 @@ type ContainerConfig struct {
 	// Command to run specified an array of strings.
 	Cmd []string `json:"Cmd"`
 
+	// The ID of the container
+	ContainerID string `json:"ContainerID,omitempty"`
```
Comment (fuweid) [inline apis/types/container_config.go]: `ContainerID` is used to be as `SandboxID`? I think it's a little confusing in the API level. Could we use annotation to handle this part?  
cc @YaoZengzeng @Starnop

File: cri/v1alpha2/cri_utils.go [hunk 6]
```diff
@@ -610,6 +616,19 @@ func applyContainerSecurityContext(lc *runtime.LinuxContainerConfig, podSandboxI
 
 // Apply Linux-specific options if applicable.
 func (c *CriManager) updateCreateConfig(createConfig *apitypes.ContainerCreateConfig, config *runtime.ContainerConfig, sandboxConfig *runtime.PodSandboxConfig, podSandboxID string) error {
+	// Apply runtime options.
+	if annotations := config.GetAnnotations(); annotations != nil {
+		createConfig.HostConfig.Runtime = annotations[anno.KubernetesRuntime]
```
Comment (fuweid) [inline cri/v1alpha2/cri_utils.go]: if the `annotations[anno.KubernetesRuntime]` is emtpy?

Comment (Starnop) [reply cri/v1alpha2/cri_utils.go]: `createConfig.HostConfig.Runtime` will be `""`, and then PouchContainer will use the `DefaultRuntime`,which is runc.
```
// set container runtime
if config.HostConfig.Runtime == "" {
	config.HostConfig.Runtime = mgr.Config.DefaultRuntime
}
```

**Comment by fuweid** [comment]
LGTM @YaoZengzeng please take a look

**Pull request merged**